server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -448,6 +451,7 @@ static final class AsyncAfterWriteAction {
                 pendingOps.incrementAndGet();
             }
             this.logger = logger;
+            this.postWriteRefreshTimeout = request instanceof BulkShardRequest bsr ? bsr.timeout : null;
```

Comment:
I think that the ctor can be changed from to `WriteRequest` to `ReplicatedWriteRequest<?>` and we will always have a timeout. Then I don't think we need null pointer handling.